### PR TITLE
#1855 Make benzene rings fused rings connection without valence errors

### LINKED ISSUE #1855: Make benzene rings fused rings connection without valence errors
**Is your feature request related to a problem? Please describe.**
Currently if add benzene ring to a structure it can create incorrect valence. See #1372 

**Describe the solution you'd like**
Following rules to be applied when add a benzene ring to existing structure

1. If connect to single atom - no changes to existing behavior
2. if connect to a single bond with two carbons, where each carbon connected to a carbon with a double bond, then rotate ring and replace double bond by single bond for the fusion
image

3. If connect to a double bond with two carbons, where each carbon connected to a carbon with a single bond, then rotate ring and use double bond for the fusion

image

4. In other cases use existing behavior

Please note that in examples above A - any atom (including no atom) and dashed bond - any bond

### PR DESCRIPTION
**Pull request by KonstantinEpam23**
Closes #1855 #1372

### CODE REVIEW

File: packages/ketcher-react/src/script/ui/action/flips.ts [hunk 1]
```diff
@@ -9,19 +9,10 @@ export function isFlipDisabled(editor): boolean {
...
-      if (bond.b.begin === atomId || bond.b.end === atomId) {
-        result.push(bondId)
-      }
-    }
-    return result
-  }
   const getBondIdsForAttachmentAtoms = () => {
     const result: number[] = []
     for (const atomId of atoms) {
-      const atomBondIds = getBondIdsForAtom(atomId)
```
Comment (KonstantinEpam23) [inline packages/ketcher-react/src/script/ui/action/flips.ts:24]: Just moved this function to atom.ts

File: packages/ketcher-core/src/application/editor/actions/template.ts [hunk 2]
```diff
@@ -277,11 +334,27 @@ function fromTemplateOnBond(restruct, template, bid, flip) {
...
+
+      pasteItems.bonds.push(newBondId)
     } else {
       const commonBond = bond.type > tmplBond.type ? bond : tmplBond
       action.mergeWith(fromBondsAttrs(restruct, existId, commonBond, true))
+
+      if (isFusingBenzeneBySpecialRules && benzeneConnectingBondType) {
+        action.addOp(
+          new BondAttr(bid, 'type', benzeneConnectingBondType).perform(restruct)
+        )
```
Comment (Konstantin1996) [inline packages/ketcher-core/src/application/editor/actions/template.ts:356]: Same here

```
this.addOption(bid, '', benzeneConnectingBondType)
```

File: packages/ketcher-core/src/application/editor/actions/template.ts [hunk 3]
```diff
@@ -277,11 +334,27 @@ function fromTemplateOnBond(restruct, template, bid, flip) {
...
-      pasteItems.bonds.push(operation.data.bid)
+      const newBondId = operation.data.bid
+
+      if (isFusingBenzeneBySpecialRules) {
+        const newBondType = benzeneDoubleBondIndexes.includes(tBondIndex)
+          ? Bond.PATTERN.TYPE.DOUBLE
+          : Bond.PATTERN.TYPE.SINGLE
+        action.addOp(
+          new BondAttr(newBondId, 'type', newBondType).perform(restruct)
+        )
```
Comment (Konstantin1996) [inline packages/ketcher-core/src/application/editor/actions/template.ts:345]: Can be moved into separate class method

```
addOption(id: number, type: string = 'type', bondType: string) {
        action.addOp(
          new BondAttr(id, type, bondType).perform(restruct)
        )
}

this.addOption(newBondId, '', newBondType)
```

**Review by Konstantin1996 — APPROVED** [review]

File: packages/ketcher-core/src/application/editor/actions/template.ts (hunk 2 above)
Comment (KonstantinEpam23) [reply packages/ketcher-core/src/application/editor/actions/template.ts:356]: action.addOp actually means "add operation" :)
I just think that that method belongs in Action class and dont see much benefit in creating a local method that will basically just call the other one.

**Review by Nitvex — APPROVED** [review]

**Pull request merged**